### PR TITLE
Support deduplication on multiple fields

### LINKED ISSUE #7: Support deduplication on multiple fields
The `dedupe` command should allow multiple fields to be listed for uniqueness

### PR DESCRIPTION
**Pull request by thePanz**
Closes #7

### CODE REVIEW

File: src/Command/Dedupe.php [hunk 1]
```diff
@@ -20,14 +20,18 @@ protected function configure(): void
     {
         $this
             ->setDescription('Removes duplicated messages')
-            ->addArgument('field', InputArgument::REQUIRED, 'Name of the field which identifies each message')
+            ->addArgument('field', InputArgument::REQUIRED | InputArgument::IS_ARRAY, 'Name of the field which identifies each message')
```
Comment (dbu) [inline src/Command/Dedupe.php]: s/field/field(s)/ in the description?

File: src/Command/Dedupe.php [hunk 2]
```diff
@@ -74,6 +65,49 @@ protected function execute(InputInterface $input, OutputInterface $output): void
...
+
+    /**
+     * @param string[] $fields
+     * @param bool[]   $seenValues Hashmap of value-hash => bool, to keep track of values already encountered
+     */
+    private function handleDuplicate(array $fields, array &$message, array &$seenValues, int $messageNum): ?array
+    {
+        $payload = json_decode($message['payload'], true);
+
+        // Check that all given fields exists in the payload:
```
Comment (dbu) [inline src/Command/Dedupe.php]: s/exists/exist/ (the plural has no s)

File: src/Command/Dedupe.php [hunk 3]
```diff
@@ -74,6 +65,49 @@ protected function execute(InputInterface $input, OutputInterface $output): void
...
+        foreach ($fields as $field) {
+            if (!array_key_exists($field, $payload)) {
+                throw new \UnexpectedValueException(sprintf(
+                    'Payload of message #%d does not have the required fields %s (%s)',
```
Comment (dbu) [inline src/Command/Dedupe.php:88]: i know this is getting complicated, but should we say fields can also be optional?

File: src/Command/Dedupe.php [hunk 4]
```diff
@@ -20,14 +20,18 @@ protected function configure(): void
...
     protected function execute(InputInterface $input, OutputInterface $output): void
     {
-        $field = $input->getArgument('field');
-        \assert(\is_string($field));
+        /** @var string[] $fields */
+        $fields = $input->getArgument('field');
+        \assert(\is_array($fields));
+
+        // Keep field names sorted to help the hash building
+        sort($fields);
```
Comment (dbu) [inline src/Command/Dedupe.php]: does the sorting matter? the order is stable if we use the order from the argument and not the order inside the message.

**Review by dbu — COMMENTED** [review]
cool! some minor suggestions

File: src/Command/Dedupe.php (hunk 4 above)
Comment (thePanz) [reply src/Command/Dedupe.php]: Agree, I just wanted to be extra sure about that sorting :) code removed

File: src/Command/Dedupe.php (hunk 3 above)
Comment (thePanz) [reply src/Command/Dedupe.php:88]: I'd leave this for a next MR, we could define fields as `?field` to make them optional :)

Comment (dbu) [reply src/Command/Dedupe.php:88]: agreed

File: src/Command/Dedupe.php [hunk 5]
```diff
@@ -74,6 +62,49 @@ protected function execute(InputInterface $input, OutputInterface $output): void
...
+            $values[$field] = $data[$field];
+        }
+
+        return sha1(serialize($values));
+    }
+
+    /**
+     * @param string[] $fields
+     * @param bool[]   $seenValues Hashmap of value-hash => bool, to keep track of values already encountered
+     */
```
Comment (dbu) [inline src/Command/Dedupe.php:75]: we should document the return value here i think: `@return array the $message if this is no duplicate, or null if $fields is in $seenValues`

File: src/Command/Dedupe.php [hunk 6]
```diff
@@ -74,6 +62,49 @@ protected function execute(InputInterface $input, OutputInterface $output): void
...
+    private function handleDuplicate(array $fields, array &$message, array &$seenValues, int $messageNum): ?array
```
Comment (dbu) [inline src/Command/Dedupe.php]: why pass $message by reference? we should not alter it in this method, i think

**Review by dbu — APPROVED** [review]

**Pull request merged**